### PR TITLE
fix(shell): on Windows :make does not echo

### LINKED ISSUE #12910: `:make` won't show program output
- `nvim --version`:NVIM v0.5.0-653-g948e625e1
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe /DWIN32 /D_WINDOWS /W3 /MD /Zi /O2 /Ob1 /DNDEBUG /W3 -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -DWIN32 -D_WIN32_WINNT=0x0600 -DINCLUDE_GENERATED_DECLARATIONS -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -IC:/projects/neovim/build/config -IC:/projects/neovim/src -IC:/projects/nvim-deps/usr/include -IC:/projects/neovim/build/src/nvim/auto -IC:/projects/neovim/build/include
Compilé par appveyor@APPVYR-WIN

Features: -acl +iconv +tui

- `vim -u DEFAULTS` : no difference
- Operating system/version: Windows 10/10.19041.388
- Terminal name/version:Windows Terminal / 1.2.2381.0
- `$TERM`:Windows doesn't have $TERM
- GNU make is shipped with MingW64
### Steps to reproduce using `nvim -u NORC`
Write a Makefile that calls eternal programs that prints to stdout
```make
run:
    @gcc test.c -o test
    ./test
```
contents of test.c
```c
#include<stdio.h>

int main()
{
  for (int i = 1; i < 10; i++)
  {
    for (int j = 1; j <= i; j ++)
    {
      printf(" %d ",j);
    }
    printf("\n");
  }
}
```
and run :make command in nvim

### Actual behavior

nvim won't display the output of the program.

```
".\test.c" 13L, 160C écrit(s)
:!make  >C:\Users\Ubuntu\AppData\Local\Temp\nvimMZw55j\2 2>&1

(1 sur 9) : 1
Appuyez sur ENTRÉE ou tapez une commande pour continuer
```

but if I run ``` make ``` directly in the shell, the output of the program will be displayed
worth to mention that nvim on Linux does not have this problem.
### Expected behavior

:make command displays the program's output in the nvim

```
"test.c" [dos] 13L, 160C written
:!make  2>&1| tee /tmp/nvimX452JB/2
 1
 1  2
 1  2  3
 1  2  3  4
 1  2  3  4  5
 1  2  3  4  5  6
 1  2  3  4  5  6  7
 1  2  3  4  5  6  7  8
 1  2  3  4  5  6  7  8  9

(1 of 9): 1
Press ENTER or type command to continue
```

### PR DESCRIPTION
**Pull request by 3N4N**
Fixes #12910

### CODE REVIEW

**Comment by 3N4N** [comment]
Test failure:
```
test\functional\helpers.lua:103: nvim_exec(): Vim(let):E488: Trailing characters: Out-File:Encoding'] = 'UTF8';Remove-Item -Force alias:cat,alias:echo,alias:sleep,alias:sort;'
```

Seems the value of `shellcmd` including a single quote is causing the problem.  Changing these to double quotes doesn't help either (checked with CI in my fork).

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -5235,9 +5235,11 @@ A jump table for the options with a short description can be found at |Q_op|.
 							*shell-powershell*
 	To use PowerShell: >
 		let &shell = executable('pwsh') ? 'pwsh' : 'powershell'
-		let &shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command [Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;'
-		let &shellredir = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
-		let &shellpipe = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
+		let &shellcmdflag = '-NoLogo -ExecutionPolicy RemoteSigned -Command '
+		let &shellcmdflag .= '[Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.UTF8Encoding]::new();'
```
Comment (justinmk) [inline runtime/doc/options.txt:5239]: The `.=` makes it look more complicated. It's ok if the line is a bit long.

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -5268,8 +5268,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 	The name of the temporary file can be represented by "%s" if necessary
```
Comment (justinmk) [inline runtime/doc/options.txt:5268]: Does the defaults list on line 5269 need to be updated? Looks like `src/nvim/options.lua` only has these two:

```
'shellpipe' 'sp'	string	(default "| tee" or "2>&1| tee")
```

File: test/functional/ex_cmds/make_spec.lua [hunk 3]
```diff
@@ -34,8 +34,8 @@ describe(':make', function()
       nvim('set_option', 'makeprg', testprg('shell-test'))
       local out = eval('execute("make")')
       -- Ensure there are no "shell returned X" messages between
-	  -- command and last line (indicating zero exit)
-      matches('LastExitCode%s+[(]', out)
+      -- command and last line (indicating zero exit)
+      matches('LastExitCode%s+ready [$]%s+[(]', out)
       matches('\n.*%: ready [$]', out)
```
Comment (justinmk) [inline test/functional/ex_cmds/make_spec.lua:39]: is this matches() redundant now ?

**Review by justinmk — APPROVED** [review]

File: test/functional/ex_cmds/make_spec.lua (hunk 3 above)
Comment (3N4N) [reply test/functional/ex_cmds/make_spec.lua:39]: I don't think so.  It checks that the exit code is zero.  If non-zero, this matches() will fail.

File: runtime/doc/options.txt (hunk 2 above)
Comment (3N4N) [reply runtime/doc/options.txt:5268]: I am confused as to how `options.lua` works (by which I mean I haven't looked into it at all).  It says, if non-windows, shellpipe will be "| tee", but in Linux with bash, I get shellpipe "2>&1| tee".  That's why I didn't know if I should change the default values in the doc.

What I _can_ say is we can remove the old cmd default: ">%s 2>&1".  But I don't know if we can remove ">", "|& tee".  Should I remove only the old-cmd default and keep the rest? or remove all but the two listed in `options.lua`?

Comment (justinmk) [reply runtime/doc/options.txt:5268]: Oh I see we have special handling for 'shellpipe': https://github.com/neovim/neovim/blob/86c164d2a14382adfb388a66699828df6c623102/src/nvim/option.c#L629-L643

> Should I remove only the previous-cmd default and keep the rest?

Yes

**Comment by T-727** [comment]
Hey! (wasn't sure if I should make an issue, if this's just noise please ignore lol)
I stumbled upon this PR and tried porting it over to my `0.8.3` config
and noticed that it makes commands like `:make` & `:grep` error if they don't generate output
(e.g if `'makeprg'` is set to only compile not run the program or grep didn't find matches).

#### Why?
`Tee-Object` doesn't create a file if it doesn't receive anything
and so nvim tries to open an errorfile which doesn't exist.

#### Solution?
Current I just added `-End { $null }` to the foreach loop (this part: `%{ }`)
which just adds null to the *end* of the results wether the loop ran or not
thus making `Tee-Object` generate an empty file when there's no output.
A bit hacky though, the best solution in my mind would be that nvim simply returns an empty quickfix list if the errorfile doesn't exist, but I'm not sure if that has any *consequences*.

Anyhow, thanks for your time 🙏

**Comment by 3N4N** [comment]
> Current I just added -End { $null } to the foreach loop (this part: %{ })
which just adds null to the end of the results wether the loop ran or not
thus making Tee-Object generate an empty file when there's no output.

Sounds fine.  Please make a PR.  I can look comment on it then.

**Pull request merged**